mlflow/openai/__init__.py

```diff
@@ -437,37 +495,52 @@ def __init__(self, model):
             )
         self.model = model
 
+    def format_messages(self, messages, params_list):
+        messages = [_Message(m) for m in messages]
+        return [[m.format(**params) for m in messages] for params in params_list]
+
+    def get_params_list(self, data, variables):
+        if len(variables) == 1:
+            variable = variables[0]
+            if variable in data.columns:
+                return data[[variable]].to_dict(orient="records")
+            else:
+                iter_string_columns = (c for c, v in data.iloc[0].items() if isinstance(v, str))
+                first_string_column = next(iter_string_columns, None)
+                if first_string_column is None:
+                    raise mlflow.MlflowException.invalid_parameter_value(
+                        "No string columns found in the input data"
+                    )
+                return [{variable: s} for s in data[first_string_column]]
+        else:
+            return data[variables].to_dict(orient="records")
+
     def predict(self, data):
-        import pandas as pd
         from mlflow.openai.api_request_parallel_processor import process_api_requests
 
-        if isinstance(data, pd.DataFrame):
-            if "content" not in data or "role" not in data:
-                raise mlflow.MlflowException.invalid_parameter_value(
-                    "Input dataframe must contain columns 'content' and 'role'",
-                )
-            messages = data.to_dict(orient="records")
-        elif isinstance(data, list) and all(isinstance(d, dict) for d in data):
-            if not all(map(_has_content_and_role, data)):
+        messages = self.model.get("messages", [])
+        variables = _collect_variables(messages)
+        if variables:
+            messages_list = self.format_messages(messages, self.get_params_list(data, variables))
+        else:
+            iter_string_columns = (c for c, v in data.iloc[0].items() if isinstance(v, str))
+            first_string_column = next(iter_string_columns, None)
+            if first_string_column is None:
                 raise mlflow.MlflowException.invalid_parameter_value(
-                    "Input list of dictionaries must contain keys 'content' and 'role'",
+                    "No string columns found in the input data"
```

Comment:
Could we make this message more clear with an example of the supported input types? "You must supply either <type x> or <type y> in order to use this API for inference"

Reply:
got it, I'll update the message :)